README
TagReader is a perl extension module which allows you to read html/xml
files by tag. That is: in a similar way as you can read textfiles by
line with "while(<>)" you use TagReader::getbytoken to read a file by tag.

To install run:
perl Makefile.PL
make
make test
make install

If you want to install TagReader and the application programs
below /usr/local (/usr/local/bin , /usr/local/lib/5.6.0/i386-linux 
/usr/local/lib/site_perl/5.6.0/i386-linux/auto/HTML/TagReader etc...) 
then run
perl Makefile.PL PREFIX=/usr/local
make
make test
make install


------------------- ------------------- -------------------
The subdirectory bin contains 4 applications of TagReader.
All of them start with the prefix "tr_"
tr_blck -- check for broken relative links in html pages
tr_llnk -- list links in html files
tr_xlnk -- expand links on directories
tr_mvlnk -- modify tags in html files with perl commands.
tr_staticssi -- expand SSI directives #include virtual and #exec cmd

tr_httpcheck -- check if a particular web-pages exists
             httpcheck does not directyl use the TagReader module
             but may be used as post processor for blck   
             If you are interessted in a link checker to check
             links only via the web-server then this is not the
             right program for you. Other programs like e.g 

              http://linkchecker.sourceforge.net/ or
	      http://www.linklint.org/ or
              http://linkchecker.stacken.kth.se/ (webpage where
              you can enter a url to check) or
	      http://htcheck.sourceforge.net/ or
	      http://www.jmarshall.com/tools/cl/

             can be used if you want to check your web-pages only 
             remotely via a web server. 

Note the primary desing goal of TagReader is to provide a fast
way of reading/processing html files. 

These application programs will normally be installed to /usr/bin/

------------------- ------------------- -------------------
Author: guido(at)linuxfocus.org

Homepage: http://linuxfocus.org/~guido/

Copyright:  This program is free software; you can redistribute it 
and/or modify it under the same terms as Perl itself.

Installation requirements: Perl and a c-compiler (e.g gcc)
------------------- ------------------- -------------------
NAME
    TagReader - Perl extension module for processing html/sgml/xml files
    by tag.

SYNOPSIS
      use HTML::TagReader;
      # open then file and get an obj-ref:
      my $p=new HTML::TagReader "filename";

      # set to zero or undef to omit warnings about html error:
      $showerrors=1; 

      # get only the tags:
      my $tag = $p->gettag($showerrors);
        # or
      my ($tag,$linenumber) = $p->gettag($showerrors);

      # get the entire file split into tags and text parts:
      my $tag = $p->getbytoken($showerrors);
        # or
      my ($tag,$tagtype,$linenumber) = $p->getbytoken($showerrors);

DESCRIPTION
    The module implements a fast and small object oriented way of processing
    any kind of html/sgml/xml files by tag.

    The getbytoken(0) is similar to while(<>) but instead of reading lines
    it reads tags or tags and text.

    Here is a program that list all href tags in a html file together with
    it line numbers:

            use TagReader;
            my $p=new TagReader "file.html";
            my @tag;
            while(@tag = $p->gettag(1)){
                    if ($tag[0]=~/ href ?=/i){
                            # remove optional space before the equal sign:
                            $tag[0]=~s/ ?= ?/=/g;
                            print "line: $tag[1]: $tag[0]\n";
                    }
            }

    Here is a program that will read a html file tag wise:

            use TagReader;
            my $p=new TagReader "file.html";
            my @tag;
            while(@tag = $p->getbytoken(1)){
                    if ($tag[1] eq ""){
                            print "line: $tag[2]: not a tag (some text), \"$tag[0]\"\n\n";
                    }else{
                            print "line: $tag[2]: is a tag, $tag[0]\n\n";
                    }
            }

  new HTML::TagReader $file;

    Returns a reference to a TagReader object. This reference can be used
    with gettag() or getbytoken() to read the next tag.

  gettag($showerrors);

    Returns in an array context tag and line number. In a scalar context
    just the next tag.

    An empty string or and empty array is returned if the file contains no
    further tags. html/xml comments and any tags inside the comments are
    ignored.

    The returned tag string has all white space (tab, newline...) reduced to
    just a single space otherwise upper and lower case, quotes etc are as in
    the original file. The line numbers are those where the tag starts.

    You must provide 0 (or undef) or 1 as an argument to gettag. If 0 is
    provided then gettag will not print any errors if it finds a syntax
    error in the html/sgml/xml code.

    Currently only the following warning messages are implemented to warn
    about possible html syntax errors:

    - A starting '<' was found but no closing '>' after 300 characters

    - A single '<' was found which was not followed by [!/a-zA-Z]. Such a
    '<' should be written as &lt;

    - A single '>' was found outside a tag.

  getbytoken($showerrors);

    Returns in an array context tag, tagtype (a, br, img,...) and line
    number. In a scalar context just the next tag.

    An empty string or and empty array is returned if the file contains no
    further tags.

    getbytoken() should be used to process a html file and possibly modify
    tags. As opposed to gettag() the getbytoken() does not remove newline or
    space from the data.

    tagtype is always lower case. The tagtype is the string starting the tag
    such as "a" in <a href=""> or "!--" in <!-- comment -->. tagtype is
    empty if this is not a tag (normal text or newline).

    You must provide 0 (or undef) or 1 as an argument to getbytoken. If 0 is
    provided then gettag will not print any errors if it finds a syntax
    error in the html/sgml/xml code.

    Currently only the following warning messages are implemented to warn
    about possible html syntax errors:

    - A starting '<' was found but no closing '>' after 300 characters

    - A single '<' was found which was not followed by [!/a-zA-Z]. Such a
    '<' should be written as &lt;

    - A single '>' was found outside a tag.

  Limitations

    There are no limitation to the size of the file.

    If you need a more sophisticated interface you might want to take a look
    at HTML::Parser. HTML:TagReader is fast generic and straight forward to
    use.

COPYRIGHT
    Copyright (c) Guido Socher [guido(at)linuxfocus.org]

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    homepage of this program: http://linuxfocus.org/~guido/

    perl(1) HTML::Parser(3)



Changes
Revision history for Perl extension HTML::TagReader.

0.01  Tue Oct  2 23:38:11 2001
	- original version; created by h2xs 1.19

0.02  Sun Jul 7 15:55:51 CEST 2002
       - added blck as example program

0.03  Sun Jul 14 16:59:41 CEST 2002
       - first public version

0.04  Mon Jul 15 18:55:10 MET DST 2002
       - changed name space to HTML::Tagreader

0.05  Fri Jul 19 17:05:10 MET DST 2002
       - documentation updates
       - bug fix: blck tries to check abs link with anchor
       - function signatures changed: No more optional arguments.

0.06  Fri Jul 19 18:10:39 CEST 2002
      - name changed from HTML::Tagreader to HTML::TagReader

0.07  Sat Jul 20 22:52:11 CEST 2002
      - bug fixes and changes in the applications llnk and xlnk

0.08  Sun Jul 21 22:59:32 CEST 2002
     - bug fix xlnk, option -v0 does not work
     - bug fix blck, <a name=xx href=...> is a valid tag.
     - unlimited file length
     - TagReader.xs re-written to use PerlIO

0.09 Mon Jul 22 22:02:18 CEST 2002
     - mvlnk added

0.10 Sun Aug 11 22:49:13 CEST 2002
     - renamed included programs to tr_something
       e.g tr_mvlnk
     - added tr_staticssi

0.11 Sun Sep  1 21:53:30 CEST 2002
     - tr_blck add option -d for checking of abs file system links
     - tr_llnk cleanup path components with ".." xx/../yy -> yy

0.12 Mon Sep  2 19:51:16 MET DST 2002
     - tr_llnk function flattenpath changed
     - tr_httpcheck be more tollerant about the input
     - do not remove // after things like http://...

0.13 Sun Oct  6 09:55:55 CEST 2002
     -"Warning, single '>' should be..." was printed even if showerrors
      was not set.
     - tr_tagcontentgrep added
     - logo added (html_trbutton_small.png)

0.14 Sun Oct  6 15:22:49 CEST 2002 
     - tr_tagcontentgrep was missing in the package

0.15 Fri Oct 18 23:19:08 CEST 2002
     - tr_fixltgt added

==========
0.50 Wed Oct 23 22:15:09 CEST 2002
     - added tr_imgaddsize
     - Interface change: In an array context one more element is
       returned. This is the column position on the line where the
       Tag starts.

0.51 Fri Oct 25 21:08:57 MET DST 2002
     - IO:Handle requires that the file descriptor is flushed before closing
       otherwise you can loose data when writing. Somehow the close does 
       not seem to flush correctly. I don't know if this is a bug in perl.

0.52 Mon Fri Dec  6 21:58:19 CET 2002
     - check in makefile for Image::Size
     - help text of tr_staticssi updated
     - tr_staticssi: -d optional

0.53 Wed Feb 26 20:45:43 CET 2003
     - updated readme
     - update tr_mvllnk, do not die in case of syntax error in expression

1.00 Fri Mar  7 19:35:04 MET 2003
     This is now a pretty stable software so we can set it to
     version 1.00
     - updated tr_blck: incorrect handling of empty links
     - disable use of locale
     - updated TagReader.xs for c++ compatibility
     - pod sections updated to fix syntax errors

1.01 Thu Jul  3 21:46:21 CEST 2003
     - tr_delfont added

1.02 Sat Jul 19 15:38:35 CEST 2003
     - wrong tagtype if tag contained numbers such as e.g <h2>

1.03 Sun Jul 20 15:29:19 CEST 2003
     - more test cases

1.04 Sun Aug  3 04:23:40 CEST 2003
     - make it compile under windows

1.05 Sat Aug 16 04:37:04 CEST 2003
     - TagReader.xs updated to print warning about too long tag only once
     - spelling errors in sample scripts updated
     - new flag for verbosity introduced in sample scripts

1.06 Tue Oct  8 06:14:14 CEST 2003
    - tr_delfont updated
    - bugfix in tr_imgaddsize

1.07 Sat Jul 31 23:45:41 CEST 2004
    - ERROR: filename must be a string scalar at ... was printed in some cases even
      if a scalar was provided.

1.08 Sun Aug 22 19:01:44 CEST 2004
    - add url decoding as suggested by Gerrit Renker

1.09 Sun Oct  3 17:29:51 CEST 2004
    - improved flattenpath subroutine in tr_llnk
    - setenv QUERY_STRING in tr_staticssi

1.10 Wed Dec  7 19:25:06 CET 2005
    - man-page editorial update
# end of changes


TagReader.pm
package HTML::TagReader;

use strict;
use vars qw($VERSION @ISA);

require DynaLoader;

@ISA = qw(DynaLoader);
$VERSION = '1.10';

bootstrap HTML::TagReader $VERSION;

1;
__END__
# Below is the documentation for your module. 

=head1 NAME

TagReader - Perl extension module for reading html/sgml/xml files
by tags.

=head1 SYNOPSIS

 use HTML::TagReader;
 # open then file and get an obj-ref:
 my $p=new HTML::TagReader "filename";

 # set to zero or undef to omit warnings about html error:
 $showerr=1; 

 # get only the tags:
 my $tag = $p->gettag($showerr);
   # or
 my ($tag,$linenumber,$column)=$p->gettag($showerr);

 # get the entire file split into tags and text parts:
 my $tagOrText = $p->getbytoken($showerr);
   # or
 my ($tagOrText,$tagtype,$linenumber,$column)=$p->getbytoken($showerr);

 # get the version of HTML::TagReader:
 my $ver=$HTML::TagReader::VERSION;

=head1 DESCRIPTION

The module implements a fast and small object oriented way of
processing any kind of html/sgml/xml files by tag.

The getbytoken(0) is similar to while(<>) but instead of reading lines 
it reads tags or tags and text. 

HTML::TagReader makes it easy to keep track of the line number in a file
even though you are not reading the file by line. This important if you
want to implement error messages about html errors in your code.

Here is a program that list all href tags
in a html file together with line numbers and column:

    use HTML::TagReader;
    my $p=new HTML::TagReader "file.html";
    my @tag;
    while(@tag = $p->gettag(1)){
            if ($tag[0]=~/ href ?=/i){
                    # remove optional space before the equal sign:
                    $tag[0]=~s/ ?= ?/=/g;
                    print "line: $tag[1]: col: $tag[2]: $tag[0]\n";
            }
    }

Here is a program that will read a html file tag
wise:

    use HTML::TagReader;
    my $p=new HTML::TagReader "file.html";
    my @tag;
    while(@tag = $p->getbytoken(1)){
            if ($tag[1] eq ""){
                    print "line: $tag[2]: col: $tag[2]: not a tag (some text), \"$tag[0]\"\n\n";
            }else{
                    print "line: $tag[2]: col: $tag[2]: is a tag, $tag[0]\n\n";
            }
    }

=head2 new HTML::TagReader $file;

Returns a reference to a TagReader object. This reference can
be used with gettag() or getbytoken() to read the next tag.
You might want to test beforehand if the file is readable and
produce your own error message if the file can not be read.
The default HTML::TagReader behavior is to die with "ERROR: Can not 
read file...".

=head2 gettag($showerr);

Returns in an array context tag, line number and character position in the
line (column). In a scalar context just the next tag is returned.
 
An empty string or and empty array is returned if the file contains
no further tags. HTML/XML comments and any tags inside the comments
are ignored.

The returned tag string has all white space (tab, newline...) reduced to just a
single space otherwise upper and lower case, quotes etc are as in the
original file. The line numbers are those where the tags
start.

You must provide 0 (or undef) or 1 as an argument to gettag. 
If 0 is provided then gettag will not print warnings if it finds
a syntax error in the html/sgml/xml code.

Currently only the following warning messages are implemented to
warn about possible html syntax errors:

- A starting '<' was found but no closing '>' after 300 characters

- A single '<' was found which was not followed by [!/a-zA-Z]. Such
a '<' should be written as &lt;

- A single '>' was found outside a tag.

=head2 getbytoken($showerr);

Returns in an array context tag, tagtype (a, br, img,...), line number
and the character position (column) in the line where the tag starts. 
In a scalar context just the next tag is returned.

An empty string or and empty array is returned if the file contains
no further tags. 

getbytoken() should be used to process a HTML file and possibly
modify tags. As opposed to gettag() the getbytoken() does not
remove newline or space from the data. getbytoken() gives you 
access to the entire file and not only to the tags. 
That is: you can process the tags and the text between the tags. 

$tagtype is always lower case. The $tagtype is the string starting
the tag such as "a" in <a href=""> or "!--" in <!-- comment -->.
$tagtype is empty if this is not a tag (normal text or newline).

You must provide 0 (or undef) or 1 as an argument to getbytoken. 
If 0 is provided then getbytoken will not print any warnings if it finds
a syntax error in the html/sgml/xml code.

Currently only the following warning messages are implemented to
warn about possible html syntax errors:

- A starting '<' was found but no closing '>' after 300 characters

- A single '<' was found which was not followed by [!/a-zA-Z]. Such
a '<' should be written as &lt;

- A single '>' was found outside a tag.

=head2 Working without HTML::TagReader

In special cases it is possible to do processing of files by tag in an
efficient way without the HTML::TagReader package. This can be done by
setting the record separator variable in perl ($/). This causes however
problems with faulty HTML code where individual '<'-characters appear in
the middle of the text. An example of such a program written in plain perl
(without HTML::TagReader) is the tr_tagcontentgrep program which is part
of the HTML::TagReader distribution. Think first then write your code!
(HTML::TagReader is in most cases the best choice, not in all ;-)

=head2 Limitations

There are no limitation to the size of the file.

If you need a more sophisticated interface you might want to take a look at
HTML::Parser. HTML:TagReader is fast, generic and straight forward to use.

=head1 COPYRIGHT

Copyright (c) Guido Socher [guido(at)linuxfocus.org]

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SEE ALSO

homepage of this program: http://linuxfocus.org/~guido/ 
or http://cpan.org/authors/id/G/GU/GUS/

perl(1) HTML::Parser(3) HTML::TokeParser(3)

=cut

